### PR TITLE
Issue 4419 - Warn users of skipped entries during ldif2db online import

### LINKED ISSUE #4419: ldif2db - online. Warn the user of skipped entries
**Issue Description**
ldif2db is used to import directory entries from an ldif file to a directory server instance. During the import operation, entries might not conform to various constraints (attribute syntax, schema violation, etc) and be skipped. On completion of the import operation the user receives a success message, with no indication of a skipped entry, apart from a message in the error logs. 

**Package Version and Platform:**
 - Platform: 5.6.6-300.fc32.x86_64
 - Package and version: 389-Directory/1.4.4.4

**Steps to Reproduce**
1. Verify syntax checking is enabled (default)
2. Generate sample ldif file
3. Modify the ldif file to violate a syntax checking rule (An empty fax number)
4. Import ldif file, notice import success message
5. Check the server error logs:
```c
ERR - slapi_entry_syntax_check - "uid=OJeleniewski1000,ou=product development,dc=example,dc=com": (facsimileTelephoneNumber) value #0 invalid per syntax
WARN - import_producer - import userroot: Skipping entry "uid=OJeleniewski1000,ou=product development,dc=example,dc=com" which violates attribute syntax,
```

**Expected results**
The user should get a warning message saying some entries were skipped and the that they should check the logs for more info.

**Additional context**
Parent bug: https://bugzilla.redhat.com/show_bug.cgi?id=1627645
ldif2db can operate in both offline and online modes, the parent bug will be divided into offline and online updates that each will require a different approach.

### PR DESCRIPTION
**Pull request by jchapma**
Bug Description:  During an online ldif2db import entries that do not
                  conform to various constraints will be skipped and
                  not imported. On completition of an import with skipped
                  entries, the server responds with a success message
                  and logs the skipped entry detail to the error logs.
                  The success messgae could lead the user to believe
                  that all entries were successfully imported.

Fix Description:  If a skipped entry occurs during import, the import
                  will continue and a warning message will be displayed.
                  The schema is extended with a nsTaskWarning attribute
                  which is used to capture and retrieve any task
                  warnings.

                  CLI tools for online import updated.

                  Test added to generate an incorrect ldif entry and perform an
                  online import.

Fixes: https://github.com/389ds/389-ds-base/issues/4419

Reviewed by: tbordaz, mreynolds389  (Thanks)

### CODE REVIEW

File: dirsrvtests/tests/suites/import/import_test.py [hunk 1]
```diff
@@ -246,7 +277,7 @@ def test_ldif2db_allows_entries_without_a_parent_to_be_imported(topo, _import_cl
     topo.standalone.start()
 
 
-def test_ldif2db_syntax_check(topo):
+def test_ldif2db_syntax_check(topo, _import_clean):
```
Comment (jchapma) [inline dirsrvtests/tests/suites/import/import_test.py:282]: Added _import_clean  function to clean up generated syntax error ldif file used for ldif2db offline test.

**Comment by droideck** [comment]
I see that the test case fails with this PR:

```
    def test_fast_slow_import(topo, _toggle_private_import_mem, _import_clean):
...
        # Let's set nsslapd-db-private-import-mem:on, nsslapd-import-cache-autosize: 0
        config = LDBMConfig(topo.standalone)
        # Measure offline import time duration total_time1
        total_time1 = _import_offline(topo, 20)
        # Now nsslapd-db-private-import-mem:off
        config.replace('nsslapd-db-private-import-mem', 'off')
        accounts = Accounts(topo.standalone, DEFAULT_SUFFIX)
        for i in accounts.filter('(uid=*)'):
            UserAccount(topo.standalone, i.dn).delete()
        # Measure offline import time duration total_time2
        total_time2 = _import_offline(topo, 20)
        # total_time1 < total_time2
>       assert total_time1 < total_time2
E       assert 2.4483795166015625 < 2.287022590637207
```

It is run without ASAN and I've repeated the test suite a few times on my VM and it was 100% reproducible. 
389-ds-base: 2.0.1-20201207git33d4ffae7.fc32
py.test -s -v dirsrvtests/tests/suites/import/import_test.py

**Comment by mreynolds389** [comment]
> I see that the test case fails with this PR:
>
>    def test_fast_slow_import(topo, _toggle_private_import_mem, _import_clean):> 
> ...
> 
> It is run without ASAN and I've repeated the test suite a few times on my VM and it was 100% reproducible.
> 389-ds-base: 2.0.1-20201207git33d4ffae7.fc32
> py.test -s -v dirsrvtests/tests/suites/import/import_test.py

I don't see how this PR could impact import performance, weird...

**Comment by Firstyear** [comment]
> I see that the test case fails with this PR:
> 
> ```
>     def test_fast_slow_import(topo, _toggle_private_import_mem, _import_clean):
> ...
>         # Let's set nsslapd-db-private-import-mem:on, nsslapd-import-cache-autosize: 0
>         config = LDBMConfig(topo.standalone)
>         # Measure offline import time duration total_time1
>         total_time1 = _import_offline(topo, 20)
>         # Now nsslapd-db-private-import-mem:off
>         config.replace('nsslapd-db-private-import-mem', 'off')
>         accounts = Accounts(topo.standalone, DEFAULT_SUFFIX)
>         for i in accounts.filter('(uid=*)'):
>             UserAccount(topo.standalone, i.dn).delete()
>         # Measure offline import time duration total_time2
>         total_time2 = _import_offline(topo, 20)
>         # total_time1 < total_time2
> >       assert total_time1 < total_time2
> E       assert 2.4483795166015625 < 2.287022590637207
> ```
> 
> It is run without ASAN and I've repeated the test suite a few times on my VM and it was 100% reproducible.
> 389-ds-base: 2.0.1-20201207git33d4ffae7.fc32
> py.test -s -v dirsrvtests/tests/suites/import/import_test.py

It's probably not a great idea to measure timing in tests anyway, because any other system noise *could* cause easy false negatives here. 

> I don't see how this PR could impact import performance, weird...

```
if((job) && (job->skipped)) {
```

I wonder if it's to do with this line. `job->skipped` is a counter, so the moment it's non-zero, we'll always attempt to set the `WARN_SKIPPED_IMPORT_ENTRY` even if it's already set. That involves a deref + fn call + OR + store, so if we get the warning early followed by many other entries we'll be hitting this function every iteration, even if on *this* iteration we didn't actually need to add the warning.

**Comment by jchapma** [comment]
This PR checks the job->skipped counter after the ldif file parsing has complete, this counter is accessed once per ldif import, so it shouldn't add too much overhead.

**Comment by Firstyear** [comment]
> This PR checks the job->skipped counter after the ldif file parsing has complete, this counter is accessed once per ldif import, so it shouldn't add too much overhead.

Ahh my bad, I must have misunderstood something then :)

**Comment by jchapma** [comment]
I would like to know the reason for the test time differences though, let me have a look at it .

**Comment by jchapma** [comment]
The test_fast_slow_import() test compares the duration of ldif import times when the **nsslapd-db-private-import-mem** config is enabled and disabled, the enabled setting being the most efficient. 

On my setup this test fails 40% of the time. This test uses dbgen_users() to create a random ldif file for each import, each with the same number of users, but with differing user details, family name, given name, etc. I noticed that the test fails when the ldif file used for the first import is larger than the ldif file used for the second import. I guess the overhead of parsing a larger ldif file could be the responsible for the mysterious time difference.

**Comment by Firstyear** [comment]
> The test_fast_slow_import() test compares the duration of ldif import times when the **nsslapd-db-private-import-mem** config is enabled and disabled, the enabled setting being the most efficient.
> 
> On my setup this test fails 40% of the time. This test uses dbgen_users() to create a random ldif file for each import, each with the same number of users, but with differing user details, family name, given name, etc. I noticed that the test fails when the ldif file used for the first import is larger than the ldif file used for the second import. I guess the overhead of parsing a larger ldif file could be the responsible for the mysterious time difference. Sometimes the generated ldif files differ by 400 bytes.

Would it be worth re-using the same ldif instead to make this more reliable?

**Comment by jchapma** [comment]
I reckon it would, this test is not part of this PR so I should create another PR for it I guess.

**Comment by Firstyear** [comment]
Yep, agreed. In that case is there any reason to not-merge this @droideck ?

**Review by droideck — APPROVED** [review]
Seems reasonable, you have my ack then! Thanks!:)

**Comment by jchapma** [comment]
Thanks for your feedback everyone.

**Pull request merged**